templates/terraform/resource.erb

```diff
@@ -262,16 +252,15 @@ func resource<%= resource_name -%>Read(d *schema.ResourceData, meta interface{})
 
 <%  object.gettable_properties.reject{|p| p.ignore_read }.each do |prop| -%>
 <%    if prop.flatten_object -%>
-    if v, ok := res["<%= prop.api_name -%>"].(map[string]interface{}); res["<%= prop.api_name -%>"] != nil && ok {
-<%      prop.properties.each do |p| -%>
-      if err := d.Set("<%= p.name.underscore -%>", flatten<%= resource_name -%><%= titlelize_property(prop) -%><%= titlelize_property(p) -%>(v["<%= p.api_name -%>"], d)); err != nil {
-        return fmt.Errorf("Error reading <%= object.name -%>: %s", err)
-    }
-<%      end -%>
-    } else {
-<%      prop.properties.each do |p| -%>
-      d.Set("<%= p.name.underscore -%>", nil)
-<%      end -%>
+// Terraform must set the top level schema field, but since this object contains collapsed properties
+// it's difficult to know what the top level should be. Instead we just loop over the map returned from flatten.
+    if flattenedProp := flatten<%= resource_name -%><%= titlelize_property(prop) -%>(res["<%= prop.api_name -%>"], d); flattenedProp != nil {
```

Comment:
We could use `root_properties` as the top level loop instead of `object.gettable_properties` which mean we wouldn't need to do this casting/looping. But it would likely end up with large downstream changes (potential property reordering) so if people are interested I'm happy to add it as another PR.